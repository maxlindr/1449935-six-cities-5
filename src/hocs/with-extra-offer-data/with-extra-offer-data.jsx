import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {StateNameSpace} from '../../store/reducers/root-reducer';
import {fetchReviewsList, fetchNearbyPlaces} from '../../store/api-actions';
import {ActionCreator} from '../../store/action';
import {offerPropTypes, reviewPropTypes} from '../../prop-types';
import {getUser} from '../../store/selectors';
import ErrorPage from '../../components/error-page/error-page';
import {ErrorMessage} from '../../constants';

const withExtraOfferData = (WrappedComponent) => {
  class WithExtraOfferData extends React.PureComponent {
    constructor(props) {
      super(props);

      const {offer, getReviews, getNearbyPlaces} = props;

      if (offer) {
        const offerId = offer.id;

        getReviews(offerId);
        getNearbyPlaces(offerId);
      }
    }

    componentWillUnmount() {
      this.props.reset();
    }

    render() {
      if (this.props.offer) {
        return <WrappedComponent {...this.props} offers={this.props.offers || []} />;
      } else {
        return <ErrorPage message={ErrorMessage.NOT_FOUND}/>;
      }
    }
  }

  WithExtraOfferData.propTypes = {
    offer: offerPropTypes.isRequired,
    reviews: PropTypes.arrayOf(reviewPropTypes),
    offers: PropTypes.arrayOf(offerPropTypes),
    getReviews: PropTypes.func.isRequired,
    getNearbyPlaces: PropTypes.func.isRequired,
    reset: PropTypes.func.isRequired,
  };

  return WithExtraOfferData;
};

const mapStateToProps = (state) => ({
  user: getUser(state),
  reviews: state[StateNameSpace.OFFER_PAGE].reviews,
  offers: state[StateNameSpace.OFFER_PAGE].nearbyPlaces,
});

const mapDispatchToProps = (dispatch) => ({
  getReviews(offerId) {
    dispatch(fetchReviewsList(offerId));
  },
  getNearbyPlaces(offerId) {
    dispatch(fetchNearbyPlaces(offerId));
  },
  reset() {
    dispatch(ActionCreator.setFetchedReviews(null));
    dispatch(ActionCreator.setFetchedNearbyPlaces(null));
  }
});

export default (WrapperComponent) => connect(mapStateToProps, mapDispatchToProps)(withExtraOfferData(WrapperComponent));
