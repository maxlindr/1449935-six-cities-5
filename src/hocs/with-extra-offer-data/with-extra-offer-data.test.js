import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import {withExtraOfferData} from './with-extra-offer-data';
import offer from '../../../__mocks__/offer-mock';

const MockComponent = () => <div>Mock Component</div>;

MockComponent.propTypes = {
  onUpdate: PropTypes.func.isRequired,
};

const MockComponentWrapped = withExtraOfferData(MockComponent);

test(`withExtraOfferData должен корректно отображаться`, () => {
  const tree = renderer.create((
    <MockComponentWrapped
      offerId={`1`}
      offer={offer}
      getReviews={() => {}}
      getNearbyPlaces={() => {}}
      reset={() => {}}
      getOffer={() => {}}
      changeCity={() => {}}
      updateOffer={() => {}}
    />
  )).toJSON();

  expect(tree).toMatchSnapshot();
});
