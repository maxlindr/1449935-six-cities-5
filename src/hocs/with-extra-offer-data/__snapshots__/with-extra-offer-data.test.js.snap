// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withExtraOfferData должен корректно отображаться 1`] = `
<div>
  Mock Component
</div>
`;
